include/igl/dual_contouring.cpp

```diff
@@ -0,0 +1,559 @@
+        //Eigen::RowVector3d e1 = e0;
+        //e1(o) -= step(o);
+        Eigen::RowVector3i jc(x,y,z);
+        jc(o) -= 1;
+        const RowVector3S e1 = primal(jc);
+        const Scalar f1 = f(e1);
+        return single_edge(x,y,z,o,e0,f0,e1,f1);
+      }
+      bool single_edge(
+        const int & x, 
+        const int & y, 
+        const int & z, 
+        const int & o,
+        const RowVector3S & e0,
+        const Scalar & f0,
+        const RowVector3S & e1,
+        const Scalar & f1)
+      {
+        const Scalar isovalue = 0;
+        if((f0>isovalue) == (f1>isovalue)) { return false; }
+        // Position of crossing point along edge
+        RowVector3S p;
+        Scalar t = -1;
+        if(root_finding)
+        {
+          Scalar tl = 0;
+          bool gl = f0>0;
+          Scalar tu = 1;
+          bool gu = f1>0;
+          assert(gu ^ gl);
+          int riter = 0;
+          const int max_riter = 7;
+          while(true)
+          {
+            t = 0.5*(tu + tl);
+            p = e0+t*(e1-e0);
+            riter++;
+            if(riter > max_riter) { break;}
+            const Scalar ft = f(p);
+            if( (ft>0) == gu) { tu = t; }
+            else if( (ft>0) == gl){ tl = t; }
+            else { break; }
+          }
+        }else
+        {
+          // inverse lerp
+          const Scalar delta = f1-f0;
+          if(delta == 0) { t = 0.5; }
+          t = (isovalue - f0)/delta;
+          p = e0+t*(e1-e0);
+        }
+        // insert vertex at this point to triangulate quad face
+        const typename decltype(V)::Index ev = triangles ? new_vertex() : -1;
+        if(triangles)
+        {
+          const std::lock_guard<std::mutex> lock(Vmut);
+          vV[ev] = p;
+          vcount[ev] = 1;
+          vI[ev] = Eigen::RowVector3i(-1,-1,-1);
+        }
+        // edge normal from function handle (could use grid finite
+        // differences/interpolation gradients)
+        const RowVector3S dfdx = f_grad(p);
+        // homogenous plane equation
+        const RowVector4S P = (RowVector4S()<<dfdx,-dfdx.dot(p)).finished();
+        // quadric contribution
+        const Matrix4S H = P.transpose() * P;
+        // Build quad face from dual vertices of 4 cells around this edge
+        Eigen::RowVector4i face;
+        int k = 0;
+        for(int i = -1;i<=0;i++)
+        {
+          for(int j = -1;j<=0;j++)
+          {
+            Eigen::RowVector3i kc(x,y,z);
+            kc(o)--;
+            kc((o+1)%3)+=i;
+            kc((o+2)%3)+=j;
+            const std::lock_guard<std::mutex> lock(Vmut);
+            const typename decltype(V)::Index v = sub2dual(kc);
+            vV[v] += p;
+            vcount[v]++;
+            vH[v] += H;
+            face(k++) = v;
+          }
+        }
+        {
+          const std::lock_guard<std::mutex> lock(Qmut);
+          if(triangles)
+          {
+            if(m+4 >= Q.rows()){ Q.conservativeResize(2*m+4,Q.cols()); }
```

Comment:
re:conservativeResize This is just amortizing the array allocation as it grows in size. The `2*m` ensures that this will only get called a few times (exponential growth of array --> log(size) number of calls). 

re:mutex, I agree. Using a map reduce would be better here. igl::parallel_for has the hooks to do it (and using tbb will be even easier/more elegant/less overhead; but that's a larger discussion). I didn't do any analysis yet on how much time is being wasted on the mutexs. I'd _guess_ that it's not much the root finding computation/memory access is the bottleneck. Worth looking into in the future. (If the atomic is a quick fix, then let's do it. I'm not confident on using it so I'll leave it for the future).

Reply:
> The 2*m ensures that this will only get called a few times (exponential growth of array --> log(size) number of calls).

Right. I though that because m grows linearly, the call to resize also had a linear amortized cost. My bad.